### PR TITLE
Made multi-scheduler graduated to Beta

### LINKED ISSUE #25318: Requirements for multi-scheduler to graduate to Beta and then v1
We'd like to graduate multi-scheduler to Beta in 1.4.

Beta
- [ ] Change annotation key from `scheduler.alpha.kubernetes.io/name` to `scheduler.beta.kubernetes.io/name`
- [ ] Make sure we have a rollback story or are sure we won't need to rollback earlier than when feature was first introduced

v1
- [ ] Change annotation key from `scheduler.beta.kubernetes.io/name` to `scheduler.kubernetes.io/name`

The design doc has some suggestions for "next steps" and we should incorporate any of those that we actually want to do into "Beta," but I'm not sure any of them are necessary?

ETA: Previously we had "Implement [MetadataPolicy](https://github.com/kubernetes/kubernetes/blob/master/docs/design/metadata-policy.md)" for Beta and "Use initializer pattern in MetadataPolicy" but I think we can separate that from this feature.

cc/ @HaiyangDING @bgrant0607 

ref/ #17865 #18262

### PR DESCRIPTION
**Pull request by k82cn**
fixes #25318

### CODE REVIEW

**Comment by davidopp** [comment]
Please remove the code that handles the annotation in this PR.

**Comment by k82cn** [comment]
@k8s-bot cvm gke e2e test this

**Comment by k82cn** [comment]
I'm working on the rebase; but it `update-all.sh` will also update docs unexpected, traced at https://github.com/kubernetes/kubernetes/issues/39505 .

**Comment by k82cn** [comment]
cc @davidopp , if any more comments, please let me know :).

**Comment by davidopp** [comment]
Please remove all the code that handles the annotation for scheduler name, e.g
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L496
should be reading from the field rather than the annotation.

Also make sure the tests are setting scheduler name using the new field you introduced, and not the annotation. (I don't understand how the tests passed without you changing factory.go)

**Comment by k82cn** [comment]
@davidopp , the annotation was removed as you mentioned, e.g. https://github.com/kubernetes/kubernetes/pull/38871/files#diff-77bdee21550bc698bfa654de11c64324L496 .

**Comment by k82cn** [comment]
In the PR, the scheduler name was set to default if empty.

**Comment by davidopp** [comment]
Ah OK, sorry I missed that! I will review the PR in the next couple of days.

**Review by timothysc — APPROVED** [review]
Only needs release note from the looks of it.

**Review by timothysc — CHANGES_REQUESTED** [review]
$ grep -r "scheduler[.]alpha" * | grep name 

still shows a fair number of results, in fact, I need to do some other cleanup work now too.

**Comment by k82cn** [comment]
@timothysc , thanks for your comments; the PR was updated accordingly :).

**Comment by timothysc** [comment]
@k82cn please squash to 2 commits.. 1 - generated code, 2 - everything else.

**Comment by k82cn** [comment]
@k8s-bot bazel test this

**Comment by k82cn** [comment]
@timothysc, the code was updated accordingly; PTAL: add a default test for default scheduler name; removed the UT case that with empty scheduler name.

**Review by timothysc — APPROVED** [review]

**Comment by timothysc** [comment]
Thanks for the cleanup. 

lgtm, but I'm slightly concerned that we don't have a multi-scheduler e2e.  Meaning we could be flying blind from a support perspective. 

I'll let @davidopp make final comments b/c of my comment.  I would propose that we create a multi-scheduler set of e2es in a separate PR.

**Comment by k82cn** [comment]
I think the integration test for multiple-scheduler at `scheduler_test.go:TestMultiScheduler` is ok; or do you want to also confirm the feature interaction with scheduler's priority/predicates?

**Comment by timothysc** [comment]
@k82cn needs release note..

**Comment by timothysc** [comment]
> I think the integration test for multiple-scheduler at scheduler_test.go:TestMultiScheduler is ok 

There are a bunch of edges around multiple schedulers that have not been vetted. 
 e.g. b2b binding to a host that runs out of resources... 

As of today we have a basic level stuff, I'm just saying we should open another issue around e2e tests to properly vet mutli-scheduling.  But there is no reason atm to block this.  

/lgtm

**Comment by k82cn** [comment]
> There are a bunch of edges around multiple schedulers that have not been vetted.
e.g. b2b binding to a host that runs out of resources...

Wow, agree with the example case :). e2e cases seems necessary.

> @ k82cn needs release note..

I'll rebase it today.

**Comment by k82cn** [comment]
@timothysc , the code was re-based. PTAL :).

**Comment by davidopp** [comment]
We decided to defer e2e tests when this feature went to alpha because we thought we were going to do a "next step" for this feature (an admission controller that would take a high-level policy configuration, and based on that, add the scheduler name annotations to the incoming pods), and I thought the integration test was sufficient until then. Well, we ended up not doing the next step, and I who knows if/when it will ever be done, so I think now is a very good time to do one or more e2e tests. @k82cn do you think you can do it in a follow-up PR?

As for back-to-back bindings from different schedulers, which @timothysc mentioned: the kubelet does admission control that checks for sufficient resources etc., so we are relying on that to ensure the node doesn't become overcomitted with respect to request. Unfortunately non-node-local properties aren't checked by the kubelet, but I think the only one that matters right now is non-node TopologyKey for pod anti-affinity (e.g. "don't put this pod in the same zone as any pods that match label selector X"), and we said we were probably going to disable that when it goes to beta in 1.6...

**Comment by k82cn** [comment]
@davidopp , sure; I'd like to add more e2e for this feature.
The follow-up PR will trace at #40125; please feel free to assign to me :).

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -2140,6 +2140,10 @@ type PodSpec struct {
 	// If specified, the pod's scheduling constraints
 	// +optional
 	Affinity *Affinity `json:"affinity,omitempty" protobuf:"bytes,18,opt,name=affinity"`
+	// If specified, the pod will be dispatched by specified scheduler.
+	// If not specified, the pod will be dispatched by default scheduler.
+	// +optional
+	SchedulerName string `json:"schedulername,omitempty" protobuf:"bytes,19,opt,name=schedulername"`
```
Comment (smarterclayton) [inline pkg/api/v1/types.go]: This is wrong - this should have been `schedulerName`

Please ensure this is fixed pre 1.6.0 - this is a release blocker.

**Comment by smarterclayton** [comment]
This has an incorrect JSON serialization for `schedulername` - must be `schedulerName`.  Please fix that ASAP, it may not go out in 1.6.

File: pkg/api/v1/types.go (hunk 1 above)
Comment (k82cn) [inline pkg/api/v1/types.go]: sure, I'll create a PR today :).

**Pull request merged**